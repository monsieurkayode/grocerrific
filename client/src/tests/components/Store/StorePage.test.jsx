import React from 'react';
import { shallow } from 'enzyme';
import {
  StorePage,
  mapStateToProps
} from '../../../components/Store/StorePage';
import mockGroceries from '../../__mocks__/mockGroceries';
import mockCartItems from '../../__mocks__/mockCartItems';

const setup = () => {
  const props = {
    groceries: mockGroceries,
    cartItems: [],
    isLoading: false,
    checkingOut: false,
    fetchGroceries: jest.fn(),
    removeGroceryFromCart: jest.fn(),
    addGroceryToCart: jest.fn(),
    clearGroceryCart: jest.fn(),
    checkout: jest.fn(() => Promise.resolve()),
  };

  const wrapper = shallow(<StorePage {...props} />);

  return { wrapper, props };
};

describe('StorePage Component', () => {
  const { wrapper, props } = setup();
  afterEach(() => wrapper.setState({
    displayModal: false,
    orderProcessed: false
  }));

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a ToastContainer', () => {
    expect(wrapper.find('ToastContainer').length).toBe(1);
  });

  it('should render GroceryItemList component', () => {
    expect(wrapper.find('GroceryItemList').length).toBe(1);
  });

  it('should render Cart if displayModal is true', () => {
    wrapper.setState({ displayModal: true });
    expect(wrapper.find('Cart').length).toBe(1);
  });

  describe('toggleModal method', () => {
    it('should be called when the cart icon is clicked', () => {
      const toggleModalSpy = jest.spyOn(wrapper.instance(), 'toggleModal');
      wrapper.instance().toggleModal();
      expect(toggleModalSpy).toHaveBeenCalled();
    });

    it('should clear the grocery cart if order has been processed', () => {
      wrapper.setState({ orderProcessed: true }).setProps({ cartItems: [] });
      wrapper.instance().toggleModal();
      expect(wrapper.instance().props.cartItems).toEqual([]);
    });
  });


  it('should call handleCheckout', () => {
    const handleCheckoutSpy = jest.spyOn(wrapper.instance(), 'handleCheckout');
    wrapper.instance().handleCheckout(props.cartItems);
    wrapper.instance().handleCheckout(mockCartItems);
    expect(handleCheckoutSpy).toHaveBeenCalled();
  });

  describe('mapStateToProps', () => {
    it('returns an object from store', () => {
      const connectedProps = mapStateToProps({
        allGroceries: {
          isLoading: false,
          groceries: mockGroceries
        },
        allCartItems: { cartItems: mockCartItems, checkingOut: false }
      });
      expect(connectedProps).toEqual({
        isLoading: false,
        groceries: mockGroceries,
        checkingOut: false,
        cartItems: mockCartItems
      });
    });
  });
});
