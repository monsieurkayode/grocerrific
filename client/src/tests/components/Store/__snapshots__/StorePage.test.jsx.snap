// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorePage Component renders correctly 1`] = `
<main
  className="container"
  id="store"
>
  <ToastContainer
    autoClose={5000}
    bodyClassName={null}
    className={null}
    closeButton={
      <CloseButton
        ariaLabel="close"
      />
    }
    closeOnClick={true}
    draggable={true}
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="top-right"
    progressClassName={null}
    progressStyle={null}
    rtl={false}
    style={null}
    toastClassName={null}
    transition={[Function]}
  />
  <withRouter(Header)>
    <div
      className="header__right"
    >
      <Link
        replace={false}
        to="/inventory"
      >
        <div
          className="start text__medium_bold uppercase"
        >
          Inventory
        </div>
      </Link>
      <div
        className="cart__wrapper"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      >
        <i
          className="material-icons cart"
        >
          shopping_cart
        </i>
        <div
          className="cart__count"
        >
          0
        </div>
      </div>
    </div>
  </withRouter(Header)>
  <GroceryItemList
    addToCart={[MockFunction]}
    cartItems={Array []}
    groceries={
      Array [
        Object {
          "_id": "5bf4754223602e00211d6587",
          "name": "mamador",
          "price": 1250,
          "quantity": 3,
        },
        Object {
          "_id": "5bf4754223602e00211d6588",
          "name": "sausage",
          "price": 120,
          "quantity": 12,
        },
        Object {
          "_id": "5bf4754223602e00211d6589",
          "name": "salad creme",
          "price": 1500,
          "quantity": 0,
        },
      ]
    }
    loading={false}
    removeFromCart={[MockFunction]}
  />
</main>
`;
